File: Iteration2/SourceCode/src/main/java/Transcript.java
import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.Map;
import java.util.Random;
import java.util.TreeMap;

public class Transcript {
    private ArrayList<CourseGrade> gradeBook = new ArrayList<CourseGrade>();
    private int completedCredit;
    private double gpa;

    public Transcript(ArrayList<CourseGrade> gradeBook,int completedCredit, double gpa) {
        this.setGradeBook(gradeBook);
        this.setCompletedCredit(completedCredit);
        this.setGpa(gpa);
    }



    public ArrayList<Course> getFailedCourses(){
        ArrayList<Course> failedCourses=new ArrayList<Course>();
        for(CourseGrade courseGrade: getGradeBook()){
            if(courseGrade.getLetterGrade().contains("FF") || (courseGrade.getLetterGrade().contains("FD"))){
                failedCourses.add(courseGrade.getCourse());
            }
        }
        return failedCourses;
    }

    public ArrayList<Course> getCompletedCourses(){
        ArrayList<Course> completedCourses=new ArrayList<Course>();
        for(CourseGrade courseGrade: getGradeBook()){
            if(!(courseGrade.getLetterGrade().contains("FF")) && !(courseGrade.getLetterGrade().contains("FD"))){
                completedCourses.add(courseGrade.getCourse());
            }

        }
        return completedCourses;
    }

    public ArrayList<CourseGrade> getGradeBook() {
        return gradeBook;
    }

    public void setGradeBook(ArrayList<CourseGrade> gradeBook) {
        this.gradeBook = gradeBook;
    }

    public int getCompletedCredit() {
        return completedCredit;
    }

    public void setCompletedCredit(int completedCredit) {
        this.completedCredit = completedCredit;
    }

    public double getGpa() {
        return gpa;
    }

    public void setGpa(double gpa) {
        this.gpa = gpa;
    }

    public String addCoursesToTranscript(ArrayList<Course> courseList,Map<String,String> courseMap) {

        String[] grades = {"AA","BA" ,"BB","CB", "CC","DC","DD","FD","FF"};
        ArrayList<CourseGrade> semesterGrades= new ArrayList<>();
        for (Course courseOfFor : courseList) {
            int randomNumber = (int)(Math.random()*9);
            CourseGrade cGrade = new CourseGrade(courseOfFor, grades[randomNumber]);
            this.gradeBook.add(cGrade);
            semesterGrades.add(cGrade);
            courseMap.put(courseOfFor.getName(),grades[randomNumber]);
        }
         DecimalFormat dt= new DecimalFormat("0.00");
         return dt.format(calculateSemesterGPA(semesterGrades));
    }
    public String calculateTotalGPA(){
        double gradeCoefficient=0;
        double courseCredit=0;
        double totalWeightOfCourses=0;
        double totalCreditOfCourses=0;
        double GPA=0;
        for(CourseGrade courseGrade: gradeBook){
            courseCredit=courseGrade.getCourse().getCredit();
            totalCreditOfCourses+=courseCredit;
            if(courseGrade.getLetterGrade().equals("FF")){ gradeCoefficient=0; }
            else if(courseGrade.getLetterGrade().equals("FD")){ gradeCoefficient=0.5; }
            else if(courseGrade.getLetterGrade().equals("DD")){ gradeCoefficient=1; }
            else if(courseGrade.getLetterGrade().equals("DC")){ gradeCoefficient=1.5; }
            else if(courseGrade.getLetterGrade().equals("CC")){ gradeCoefficient=2.0; }
            else if(courseGrade.getLetterGrade().equals("CB")){ gradeCoefficient=2.5; }
            else if(courseGrade.getLetterGrade().equals("BB")){ gradeCoefficient=3.0; }
            else if(courseGrade.getLetterGrade().equals("BA")){ gradeCoefficient=3.5; }
            else if(courseGrade.getLetterGrade().equals("AA")){ gradeCoefficient=4.0; }
            totalWeightOfCourses+= gradeCoefficient*courseCredit;
        }
        GPA=totalWeightOfCourses/totalCreditOfCourses;
        this.setGpa(GPA);
        DecimalFormat dt= new DecimalFormat("0.00");
        return dt.format(GPA);
    }

    public double calculateSemesterGPA(ArrayList<CourseGrade> courseGradeList){
        double gradeCoefficient=0;
        double courseCredit=0;
        double totalWeightOfSemesterCourses=0;
        double totalCreditOfSemesterCourses=0;
        double semesterGPA=0;
        for(CourseGrade courseGrade: courseGradeList){
            courseCredit=courseGrade.getCourse().getCredit();
            totalCreditOfSemesterCourses+=courseCredit;
            String courseLetterGrade=courseGrade.getLetterGrade();
            if(courseLetterGrade.equals("FF")){ gradeCoefficient=0; }
            else if(courseLetterGrade.equals("FD")){ gradeCoefficient=0.5; }
            else if(courseLetterGrade.equals("DD")){ gradeCoefficient=1; }
            else if(courseLetterGrade.equals("DC")){ gradeCoefficient=1.5; }
            else if(courseLetterGrade.equals("CC")){ gradeCoefficient=2.0; }
            else if(courseLetterGrade.equals("CB")){ gradeCoefficient=2.5; }
            else if(courseLetterGrade.equals("BB")){ gradeCoefficient=3.0; }
            else if(courseLetterGrade.equals("BA")){ gradeCoefficient=3.5; }
            else if(courseLetterGrade.equals("AA")){ gradeCoefficient=4.0; }
            totalWeightOfSemesterCourses+=gradeCoefficient*courseCredit;
        }
        semesterGPA=totalWeightOfSemesterCourses/totalCreditOfSemesterCourses;
        return semesterGPA;
    }
}

File: Iteration3/SourceCode/Transcript.py
import random

from CourseGrade import CourseGrade


class Transcript(object):

    def __init__(self, gradeBook, completedCredit, gpa):
        gradeBook = []
        completedCredit = 0
        gpa = 0

        self.setGradeBook(gradeBook)
        self.setCompletedCredit(completedCredit)
        self.setGpa(gpa)



    def getFailedCourses(self):
        failedCourses = []
        for courseGrade in self.getGradeBook():
            if "FF" in courseGrade.getLetterGrade() or ("FD" in courseGrade.getLetterGrade()):
                failedCourses.append(courseGrade.getCourse())
        return failedCourses

    def getCompletedCourses(self):
        completedCourses = []
        for courseGrade in self.getGradeBook():
            if not("FF" in courseGrade.getLetterGrade()) and not("FD" in courseGrade.getLetterGrade()):
                completedCourses.append(courseGrade.getCourse())

        return completedCourses

    def getGradeBook(self):
        return self.__gradeBook

    def setGradeBook(self, gradeBook):
        self.__gradeBook = gradeBook

    def getCompletedCredit(self):
        return self.__completedCredit

    def setCompletedCredit(self, completedCredit):
        self.__completedCredit = completedCredit

    def getGpa(self):
        return self.__gpa

    def setGpa(self, gpa):
        self.__gpa = gpa

    def addCoursesToTranscript(self, courseList, courseMap):

        grades = ["AA", "BA", "BB", "CB", "CC", "DC", "DD", "FD", "FF"]
        semesterGrades = []
        for courseOfFor in courseList:
            randomNumber = int((random.random()*9))
            cGrade = CourseGrade(courseOfFor, grades[randomNumber])
            self.__gradeBook.append(cGrade)
            semesterGrades.append(cGrade)
            courseMap.update({courseOfFor.getName(): grades[randomNumber]})
        #dt = java.text.DecimalFormat("0.00")
        return format(self.calculateSemesterGPA(semesterGrades))
    def calculateTotalGPA(self):
        gradeCoefficient =0
        courseCredit =0
        totalWeightOfCourses =0
        totalCreditOfCourses =1 #değiştirilcek
        GPA =0
        for courseGrade in self.__gradeBook:
            courseCredit=courseGrade.getCourse().getCredit()
            totalCreditOfCourses+=courseCredit
            if courseGrade.getLetterGrade() == "FF":
                gradeCoefficient=0
            elif courseGrade.getLetterGrade() == "FD":
                gradeCoefficient=0.5
            elif courseGrade.getLetterGrade() == "DD":
                gradeCoefficient=1
            elif courseGrade.getLetterGrade() == "DC":
                gradeCoefficient=1.5
            elif courseGrade.getLetterGrade() == "CC":
                gradeCoefficient=2.0
            elif courseGrade.getLetterGrade() == "CB":
                gradeCoefficient=2.5
            elif courseGrade.getLetterGrade() == "BB":
                gradeCoefficient=3.0
            elif courseGrade.getLetterGrade() == "BA":
                gradeCoefficient=3.5
            elif courseGrade.getLetterGrade() == "AA":
                gradeCoefficient=4.0
            totalWeightOfCourses+= gradeCoefficient*courseCredit
        GPA=totalWeightOfCourses/totalCreditOfCourses
        self.setGpa(GPA)
        #dt = java.text.DecimalFormat("0.00")
        return format(GPA)

    def calculateSemesterGPA(self, courseGradeList):
        gradeCoefficient =0
        courseCredit =0
        totalWeightOfSemesterCourses =0
        totalCreditOfSemesterCourses =0
        semesterGPA =0
        for courseGrade in courseGradeList:
            courseCredit=courseGrade.getCourse().getCredit()
            totalCreditOfSemesterCourses+=courseCredit
            courseLetterGrade =courseGrade.getLetterGrade()
            if courseLetterGrade == "FF":
                gradeCoefficient=0
            elif courseLetterGrade == "FD":
                gradeCoefficient=0.5
            elif courseLetterGrade == "DD":
                gradeCoefficient=1
            elif courseLetterGrade == "DC":
                gradeCoefficient=1.5
            elif courseLetterGrade == "CC":
                gradeCoefficient=2.0
            elif courseLetterGrade == "CB":
                gradeCoefficient=2.5
            elif courseLetterGrade == "BB":
                gradeCoefficient=3.0